dbms/src/Storages/KVStore/MultiRaft/PrehandleSnapshot.cpp

```diff
@@ -0,0 +1,315 @@
+                    // Can not decode data with `force_decode == true`, must be something wrong
+                    throw;
+                }
+
+                // Update schema and try to decode again
+                LOG_INFO(
+                    log,
+                    "Decoding Region snapshot data meet error, sync schema and try to decode again {} [error={}]",
+                    new_region->toString(true),
+                    e.displayText());
+                GET_METRIC(tiflash_schema_trigger_count, type_raft_decode).Increment();
+                tmt.getSchemaSyncerManager()->syncTableSchema(context, keyspace_id, physical_table_id);
+                // Next time should force_decode
+                force_decode = true;
+
+                continue;
+            }
+            else if (e.code() == ErrorCodes::TABLE_IS_DROPPED)
+            {
+                // We can ignore if storage is dropped.
+                LOG_INFO(
+                    log,
+                    "Pre-handle snapshot to DTFiles is ignored because the table is dropped {}",
+                    new_region->toString(true));
+                break;
+            }
+            else
+            {
+                // Other unrecoverable error, throw
+                e.addMessage(fmt::format("keyspace={} physical_table_id={}", keyspace_id, physical_table_id));
+                throw;
+            }
+        }
+    }
+
+    return prehandle_result;
+}
+
+template <>
+void KVStore::releasePreHandledSnapshot<RegionPtrWithSnapshotFiles>(
+    const RegionPtrWithSnapshotFiles & s,
+    TMTContext & tmt)
+{
+    auto & storages = tmt.getStorages();
+    auto keyspace_id = s.base->getKeyspaceID();
+    auto table_id = s.base->getMappedTableID();
+    auto storage = storages.get(keyspace_id, table_id);
+    if (storage->engineType() != TiDB::StorageEngine::DT)
+    {
+        return;
+    }
+    auto dm_storage = std::dynamic_pointer_cast<StorageDeltaMerge>(storage);
+    LOG_INFO(
+        log,
+        "Release prehandled snapshot, clean {} dmfiles, region_id={} keyspace={} table_id={}",
+        s.external_files.size(),
+        s.base->id(),
+        keyspace_id,
+        table_id);
+    auto & context = tmt.getContext();
+    dm_storage->cleanPreIngestFiles(s.external_files, context.getSettingsRef());
+}
+
+void KVStore::abortPreHandleSnapshot(UInt64 region_id, TMTContext & tmt)
+{
+    UNUSED(tmt);
+    auto cancel_flag = prehandling_trace.deregisterTask(region_id);
+    if (cancel_flag)
+    {
+        // The task is registered, set the cancel flag to true and the generated files
+        // will be clear later by `releasePreHandleSnapshot`
+        LOG_INFO(log, "Try cancel pre-handling from upper layer, region_id={}", region_id);
+        cancel_flag->store(true, std::memory_order_seq_cst);
+    }
+    else
+    {
+        // the task is not registered, continue
+        LOG_INFO(log, "Start cancel pre-handling from upper layer, region_id={}", region_id);
+    }
+}
```

Comment:
```suggestion
void KVStore::abortPreHandleSnapshot(UInt64 region_id, TMTContext & tmt)
{
    UNUSED(tmt);
    auto cancel_flag = prehandling_trace.deregisterTask(region_id);
    if (cancel_flag)
    {
        // The task is registered, set the cancel flag to true and the generated files
        // will be clear later by `releasePreHandleSnapshot`
        LOG_INFO(log, "Try cancel pre-handling from upper layer, region_id={}", region_id);
        cancel_flag->store(true, std::memory_order_seq_cst);
    }
    else
    {
        // the task is not registered, continue
        LOG_INFO(log, "Start cancel pre-handling from upper layer, region_id={}", region_id);
    }
}

template <>
void KVStore::releasePreHandledSnapshot<RegionPtrWithSnapshotFiles>(
    const RegionPtrWithSnapshotFiles & s,
    TMTContext & tmt)
{
    auto & storages = tmt.getStorages();
    auto keyspace_id = s.base->getKeyspaceID();
    auto table_id = s.base->getMappedTableID();
    auto storage = storages.get(keyspace_id, table_id);
    if (storage->engineType() != TiDB::StorageEngine::DT)
    {
        return;
    }
    auto dm_storage = std::dynamic_pointer_cast<StorageDeltaMerge>(storage);
    LOG_INFO(
        log,
        "Release prehandled snapshot, clean {} dmfiles, region_id={} keyspace={} table_id={}",
        s.external_files.size(),
        s.base->id(),
        keyspace_id,
        table_id);
    auto & context = tmt.getContext();
    dm_storage->cleanPreIngestFiles(s.external_files, context.getSettingsRef());
}
```

Reply:
I prefer to put `abortXXX` before `releaseXXX` because they have logical sequence